src/utilities/include/svd.h

```diff
@@ -0,0 +1,310 @@
+  float rho3 = Dist2(B[2]);
+  bool c;
+  c = rho1 < rho2;
+  CondNegSwap(c, B[0][0], B[1][0]);
+  CondNegSwap(c, V[0][0], V[1][0]);
+  CondNegSwap(c, B[0][1], B[1][1]);
+  CondNegSwap(c, V[0][1], V[1][1]);
+  CondNegSwap(c, B[0][2], B[1][2]);
+  CondNegSwap(c, V[0][2], V[1][2]);
+  CondSwap(c, rho1, rho2);
+  c = rho1 < rho3;
+  CondNegSwap(c, B[0][0], B[2][0]);
+  CondNegSwap(c, V[0][0], V[2][0]);
+  CondNegSwap(c, B[0][1], B[2][1]);
+  CondNegSwap(c, V[0][1], V[2][1]);
+  CondNegSwap(c, B[0][2], B[2][2]);
+  CondNegSwap(c, V[0][2], V[2][2]);
+  CondSwap(c, rho1, rho3);
+  c = rho2 < rho3;
+  CondNegSwap(c, B[1][0], B[2][0]);
+  CondNegSwap(c, V[1][0], V[2][0]);
+  CondNegSwap(c, B[1][1], B[2][1]);
+  CondNegSwap(c, V[1][1], V[2][1]);
+  CondNegSwap(c, B[1][2], B[2][2]);
+  CondNegSwap(c, V[1][2], V[2][2]);
+}
+// Implementation of Algorithm 4
+Givens QRGivensQuaternion(float a1, float a2) {
+  // a1 = pivot point on diagonal
+  // a2 = lower triangular entry we want to annihilate
+  float epsilon = (float)_SVD_EPSILON;
+  float rho = sqrt(fmaf(a1, a1, +a2 * a2));
+  Givens g{fabsf(a1) + fmaxf(rho, epsilon), rho > epsilon ? a2 : 0};
+  bool b = a1 < 0.f;
+  CondSwap(b, g.sh, g.ch);
+  float w = 1.f / sqrt(fmaf(g.ch, g.ch, g.sh * g.sh));
+  g.ch *= w;
+  g.sh *= w;
+  return g;
+}
+// Implements a QR decomposition of a Matrix, see Sec 4.2
+QR QRDecomposition(glm::mat3& B) {
+  glm::mat3 Q, R;
+  // first Givens rotation (ch,0,0,sh)
+  auto g1 = QRGivensQuaternion(B[0][0], B[0][1]);
+  auto a = fmaf(-2.f, g1.sh * g1.sh, 1.f);
+  auto b = 2.f * g1.ch * g1.sh;
+  // apply B = Q' * B
+  R[0][0] = fmaf(a, B[0][0], b * B[0][1]);
+  R[1][0] = fmaf(a, B[1][0], b * B[1][1]);
+  R[2][0] = fmaf(a, B[2][0], b * B[2][1]);
+  R[0][1] = fmaf(-b, B[0][0], a * B[0][1]);
+  R[1][1] = fmaf(-b, B[1][0], a * B[1][1]);
+  R[2][1] = fmaf(-b, B[2][0], a * B[2][1]);
+  R[0][2] = B[0][2];
+  R[1][2] = B[1][2];
+  R[2][2] = B[2][2];
+  // second Givens rotation (ch,0,-sh,0)
+  auto g2 = QRGivensQuaternion(R[0][0], R[0][2]);
+  a = fmaf(-2.f, g2.sh * g2.sh, 1.f);
+  b = 2.f * g2.ch * g2.sh;
+  // apply B = Q' * B;
+  B[0][0] = fmaf(a, R[0][0], b * R[0][2]);
+  B[1][0] = fmaf(a, R[1][0], b * R[1][2]);
+  B[2][0] = fmaf(a, R[2][0], b * R[2][2]);
+  B[0][1] = R[0][1];
+  B[1][1] = R[1][1];
+  B[2][1] = R[2][1];
+  B[0][2] = fmaf(-b, R[0][0], a * R[0][2]);
+  B[1][2] = fmaf(-b, R[1][0], a * R[1][2]);
+  B[2][2] = fmaf(-b, R[2][0], a * R[2][2]);
+  // third Givens rotation (ch,sh,0,0)
+  auto g3 = QRGivensQuaternion(B[1][1], B[1][2]);
+  a = fmaf(-2.f, g3.sh * g3.sh, 1.f);
+  b = 2.f * g3.ch * g3.sh;
+  // R is now set to desired value
+  R[0][0] = B[0][0];
+  R[1][0] = B[1][0];
+  R[2][0] = B[2][0];
+  R[0][1] = fmaf(a, B[0][1], b * B[0][2]);
+  R[1][1] = fmaf(a, B[1][1], b * B[1][2]);
+  R[2][1] = fmaf(a, B[2][1], b * B[2][2]);
+  R[0][2] = fmaf(-b, B[0][1], a * B[0][2]);
+  R[1][2] = fmaf(-b, B[1][1], a * B[1][2]);
+  R[2][2] = fmaf(-b, B[2][1], a * B[2][2]);
```

Comment:
I could simplify this stuff too, using GLM vector ops, but since it's operating on rows instead of columns, I'd need to transpose and I decided it was simpler to just leave it for now. And who knows, maybe this really is slightly faster?